frontend/src/pages/new_question/utils/queryBuilderUtils.js

```diff
@@ -0,0 +1,199 @@
+ * @param {string} rootNode root node of graph
+ */
+function trimDetached(q_graph, rootNode) {
+  // all edges start out as disconnected and we'll remove them when we find a connection to root
+  const disconnectedEdges = new Set(Object.keys(q_graph.edges));
+  // we add to connected nodes, starting default with the root node
+  const connectedNodes = new Set([rootNode]);
+  let foundConnections = true;
+  // traverse graph and delete all edges not connected
+  while (foundConnections) {
+    let foundConnection = false;
+    // loop over all connected nodes to find edges that are connected to them
+    [...connectedNodes].forEach((nodeId) => {
+      [...disconnectedEdges].forEach((e) => {
+        const edge = q_graph.edges[e];
+        // if edge is connected to node
+        if (edge.subject === nodeId || edge.object === nodeId) {
+          connectedNodes.add(edge.subject);
+          connectedNodes.add(edge.object);
+          disconnectedEdges.delete(e);
+          foundConnection = true;
+        }
+      });
+    });
+    // no more connected edges, break out
+    if (!foundConnection) {
+      foundConnections = false;
+    }
+  }
+  q_graph.edges = _.omitBy(q_graph.edges, (e, id) => disconnectedEdges.has(id));
+  // delete all floating nodes
+  const notFloatingNodeIDs = new Set();
+  Object.values(q_graph.edges).forEach((e) => {
+    notFloatingNodeIDs.add(e.subject);
+    notFloatingNodeIDs.add(e.object);
+  });
+
+  // Trim a node if it is floating and marked for deletion
+  const nodesToTrim = Object.keys(q_graph.nodes).filter((id) => (!notFloatingNodeIDs.has(id) && id !== rootNode));
+
+  q_graph.nodes = _.omitBy(q_graph.nodes, (n, id) => nodesToTrim.includes(id));
+  // q_graph.nodes = _.pick(q_graph.nodes, notFloatingNodeIDs);
+  return q_graph;
+}
+
+/**
+ * After deleting a node, trim any edges connected to it
+ * @param {object} q_graph query graph
+ * @param {string} deletedNode deleted node id
+ * @param {string} rootNode root node id
+ * @returns trimDetached query graph
+ */
+function trimDetachedEdges(q_graph, deletedNode, rootNode) {
+  const edgeIds = Object.keys(q_graph.edges).map((id) => id);
+  edgeIds.forEach((eId) => {
+    const currentEdge = q_graph.edges[eId];
+    if (currentEdge.subject === deletedNode || currentEdge.object === deletedNode) {
+      delete q_graph.edges[eId];
+    }
+  });
+  let root = rootNode;
+  if (root === deletedNode) {
+    root = findRootNode(q_graph);
+  }
+  return trimDetached(q_graph, root);
+}
+
+/**
+ * Get a unique list of all edges connected to specified node
+ * @param {object} edges query graph edges object
+ * @param {string} nodeId node id
+ * @returns Set of edge ids
+ */
+function getConnectedEdges(edges, nodeId) {
+  return new Set(Object.keys(edges).filter((eId) => (
+    edges[eId].subject === nodeId || edges[eId].object === nodeId
+  )));
+}
+
+function computeRootNode(q_graph, rootNode) {
+  if (getConnectedEdges(q_graph.edges, rootNode).size > 0) {
```

Comment:
If we just don't make this check and re-find the root node always, will it work the same?

Reply:
I don't think we should be recomputing the root node any time the graph changes. Then we'd just be keeping the larger side of the graph each time, which I don't think we want. We only want to recompute the root when it is deleted.